File: FirebaseDatabase/Sources/Core/FQuerySpec.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseDatabase/Sources/Core/FQuerySpec.h"

@interface FQuerySpec ()

@property(nonatomic, strong, readwrite) FPath *path;
@property(nonatomic, strong, readwrite) FQueryParams *params;

@end

@implementation FQuerySpec

- (id)initWithPath:(FPath *)path params:(FQueryParams *)params {
    self = [super init];
    if (self != nil) {
        self->_path = path;
        self->_params = params;
    }
    return self;
}

+ (FQuerySpec *)defaultQueryAtPath:(FPath *)path {
    return [[FQuerySpec alloc] initWithPath:path
                                     params:[FQueryParams defaultInstance]];
}

- (id)copyWithZone:(NSZone *)zone {
    // Immutable
    return self;
}

- (id<FIndex>)index {
    return self.params.index;
}

- (BOOL)isDefault {
    return self.params.isDefault;
}

- (BOOL)loadsAllData {
    return self.params.loadsAllData;
}

- (BOOL)isEqual:(id)object {
    if (self == object) {
        return YES;
    }

    if (![object isKindOfClass:[FQuerySpec class]]) {
        return NO;
    }

    FQuerySpec *other = (FQuerySpec *)object;

    if (![self.path isEqual:other.path]) {
        return NO;
    }

    return [self.params isEqual:other.params];
}

- (NSUInteger)hash {
    return self.path.hash * 31 + self.params.hash;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"FQuerySpec (path: %@, params: %@)",
                                      self.path, self.params];
}

@end

File: FirebaseDatabaseSwiftCore/Sources/Core/FQuerySpec.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 03/03/2022.
//

import Foundation

@objc public class FQuerySpec: NSObject, NSCopying {
    @objc public let path: FPath
    @objc public let params: FQueryParams
    @objc public init(path: FPath, params: FQueryParams) {
        self.params = params
        self.path = path
    }

    public func copy(with zone: NSZone? = nil) -> Any {
        // Immutable
        self
    }

    @objc public static func defaultQueryAtPath(_ path: FPath) -> FQuerySpec {
        FQuerySpec(path: path, params: .defaultInstance)
    }

    @objc public var index: FIndex {
        params.index
    }
    @objc public var isDefault: Bool {
        params.isDefault
    }
    @objc public var loadsAllData: Bool {
        params.loadsAllData
    }

    @objc public override func isEqual(_ object: Any?) -> Bool {
        guard let other = object as? FQuerySpec else { return false }
        if self === other { return true }
        return self.path == other.path && self.params == other.params
    }

    @objc public override var hash: Int {
        var hasher = Hasher()
        path.hash(into: &hasher)
        params.hash(into: &hasher)
        return hasher.finalize()
    }

    @objc public override var description: String {
        "FQuerySpec (path: \(path), params: \(params)"
    }
}
